crates/matrix-sdk/src/event_cache/room/events.rs

```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
+    fn remove_events_and_update_insert_position(
+        &mut self,
+        event_ids: Vec<OwnedEventId>,
+        position: &mut Position,
+    ) {
+        for event_id in event_ids {
+            let Some(event_position) = self.revents().find_map(|(position, event)| {
+                (event.event_id().as_ref() == Some(&event_id)).then_some(position)
+            }) else {
+                error!(?event_id, "Cannot find te event to remove");
+
+                continue;
+            };
+
+            self.chunks
+                .remove_item_at(event_position)
+                .expect("Failed to remove an event we have just found");
+
+            // A `Position` is composed of a `ChunkIdentifier` and an index.
+            // The `ChunkIdentifier` is stable, i.e. it won't change if an
+            // event is removed in another chunk. It means we only need to
+            // update `position` if the removal happened in **the same
+            // chunk**.
+            if event_position.chunk_identifier() == position.chunk_identifier() {
+                // Now we can compare the position indices.
+                match event_position.index().cmp(&position.index()) {
+                    // `event_position`'s index < `position`'s index
+                    Ordering::Less => {
+                        // An event has been removed _before_ the new
+                        // events: `position` needs to be shifted to the
+                        // left by 1.
+                        position.move_index_to_the_left();
+                    }
+
+                    // `event_position`'s index >= `position`'s index
+                    Ordering::Equal | Ordering::Greater => {
+                        // An event has been removed at the _same_ position of
+                        // or _after_ the new events: `position` does _NOT_ need
+                        // tp be modified.
+                    }
+                }
+            }
+        }
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use assert_matches2::assert_let;
-    use matrix_sdk_test::{EventBuilder, ALICE};
-    use ruma::{events::room::message::RoomMessageEventContent, EventId, OwnedEventId};
+    use ruma::{user_id, EventId, OwnedEventId};
 
     use super::*;
+    use crate::test_utils::events::EventFactory;
+
+    macro_rules! assert_events_eq {
+        ( $events_iterator:expr, [ $( ( $event_id:ident at ( $chunk_identifier:literal, $index:literal ) ) ),* $(,)? ] ) => {
+            {
+                let mut events = $events_iterator;
+
+                $(
+                    assert_let!(Some((position, event)) = events.next());
+                    assert_eq!(position.chunk_identifier(), $chunk_identifier );
+                    assert_eq!(position.index(), $index );
+                    assert_eq!(event.event_id().unwrap(), $event_id );
+                )*
+
+                assert!(events.next().is_none(), "No more event is expected");
```

Comment:
Zero is plural in English :nerd_face: 

```suggestion
                assert!(events.next().is_none(), "No more events are expected");
```

Reply:
TIL